<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Genotyping by RNA Sequencing Tutorial: Preparing GBRS Pseudo-References</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="manifest" href="../site.webmanifest"><link rel="mask-icon" href="../safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav software"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Software Carpentry" src="../assets/images/software-logo.svg"><abbr class="badge badge-light" title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught." style="background-color: #FF4955; border-radius: 5px">
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="external-link alert-link" style="color: #000">
            <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="border-radius: 5px"></i>
            Pre-Alpha
          </a>
          <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../prepare-pseudo-reference.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav software" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Software Carpentry" src="../assets/images/software-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Genotyping by RNA Sequencing Tutorial
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Genotyping by RNA Sequencing Tutorial
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a class="btn btn-primary" href="../aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Genotyping by RNA Sequencing Tutorial
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 19%" class="percentage">
    19%
  </div>
  <div class="progress software">
    <div class="progress-bar software" role="progressbar" style="width: 19%" aria-valuenow="19" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../prepare-pseudo-reference.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->
      
            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Genotyping by RNA Sequencing</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="installation.html">2. Installing GBRS</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        3. Preparing GBRS Pseudo-References
        </span>
      </button>
    </div><!--/div.accordion-header-->
        
    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#introduction">Introduction</a></li>
<li><a href="#creating-a-pseudo-reference-for-diversity-outbred-mice">Creating a Pseudo-Reference for Diversity Outbred Mice</a></li>
<li><a href="#creating-pseudo-reference-genomes-and-transcriptomes-for-other-strains">Creating Pseudo-Reference Genomes and Transcriptomes for Other
Strains</a></li>
<li><a href="#step-by-step-pseudo-reference-generation">Step-by-Step Pseudo-Reference Generation</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="prepare-emase.html">4. Prepare EMASE Reference Files</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="run-gbrs.html">5. Running GBRS</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="../instructor/aio.html">See all in one page</a>
            

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">
            
            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/installation.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/prepare-emase.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/installation.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Installing GBRS
        </a>
        <a class="chapter-link float-end" href="../instructor/prepare-emase.html" rel="next">
          Next: Prepare EMASE... 
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Preparing GBRS Pseudo-References</h1>
        <p>Last updated on 2024-03-27 |
        
        <a href="https://github.com/dmgatti/GBRS_Tutorial/edit/main/episodes/prepare-pseudo-reference.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
        
        
        
        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 75 minutes</p>
        
        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>Why do we need to align to a non-reference genome?</li>
<li>What is a pseudo-reference genome?</li>
<li>How do we create a pseudo-reference genome?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Construct a pseudo-reference genome for Diversity Outbred mice.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section id="introduction"><h2 class="section-heading">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width"><p><a href="https://github.com/churchill-lab/g2gtools" class="external-link">Genome-to-Genome
Tools</a>, (g2gtools) is a suite of tools which incorporates SNPs and
indels from mouse strains into the reference genome. Inserting these
variants may lead to changes in the length of the genome and g2gtools is
able to map positions between the reference genome and the modified
genome.</p>
<p>In this tutorial, we will show you how to use the pre-built
<code>nextflow</code> pipeline to build a pseudo-reference on the JAX
computing cluster (sumner2). Then, we will show you how each step work.
We expect that most users will use the <code>nextflow</code> pipeline
and have included the step-by-step pipeline for advanced users who may
want to use non-standard pipeline options.</p>
</section><section id="creating-a-pseudo-reference-for-diversity-outbred-mice"><h2 class="section-heading">Creating a Pseudo-Reference for Diversity Outbred Mice<a class="anchor" aria-label="anchor" href="#creating-a-pseudo-reference-for-diversity-outbred-mice"></a>
</h2>
<hr class="half-width"><p>The <code>nextflow</code> pipeline developed by Computational
Sciences contains default values to make pseudo-reference generation for
Diversity Outbred mice easy. You can run the commands with minimal
arguments.</p>
<p>There is a Quick Start Guide for JAX users at <a href="https://github.com/TheJacksonLaboratory/cs-nf-pipelines/wiki/Pipeline-Documentation#quick-start-for-jax-users" class="external-link uri">https://github.com/TheJacksonLaboratory/cs-nf-pipelines/wiki/Pipeline-Documentation#quick-start-for-jax-users</a>.
This describes the essentials of running a <code>nextflow</code>
pipeline on sumner2 at JAX.</p>
<p>The documentation for the GBRS <code>nextflow</code> pipeline is at
<a href="https://github.com/TheJacksonLaboratory/cs-nf-pipelines/wiki/Generate-Pseudoreference-Pipeline-ReadMe" class="external-link uri">https://github.com/TheJacksonLaboratory/cs-nf-pipelines/wiki/Generate-Pseudoreference-Pipeline-ReadMe</a>.
This page lists the default values, which point reference files on
sumner2 and list the eight Diversity Outbred founder strains by default.
Some of the default files are listed in the table below.</p>
<table class="table"><colgroup><col width="22%"><col width="30%"><col width="22%"><col width="26%"></colgroup><thead><tr class="header"><th>File Type</th>
<th>Argument Name</th>
<th>File Path</th>
<th>Description</th>
</tr></thead><tbody><tr class="odd"><td>Reference FASTA</td>
<td>–primary_reference_fasta</td>
<td>/projects/compsci/omics_share/mouse/GRCm39/genome/sequence/ensembl/v105/Mus_musculus.GRCm39.dna.primary_assembly.fa</td>
<td>primary reference fasta file for GRCm39</td>
</tr><tr class="even"><td>SNP VCF</td>
<td>–snp_vcf</td>
<td>/projects/compsci/omics_share/mouse/GRCm39/genome/annotation/snps_indels/rel_2112_v8/mgp_REL2021_snps.vcf.gz</td>
<td>VCF containing SNPs in mouse strains from the <a href="https://www.mousegenomes.org/" class="external-link">Mouse Genomes Project</a> version
8</td>
</tr><tr class="odd"><td>Indel VCF</td>
<td>–indel_vcf</td>
<td>/projects/compsci/omics_share/mouse/GRCm39/genome/annotation/snps_indels/rel_2112_v8/mgp_REL2021_indels.vcf.gz</td>
<td>VCF containing Indels in mouse strains from the <a href="https://www.mousegenomes.org/" class="external-link">Mouse Genomes Project</a> version
8</td>
</tr><tr class="even"><td>Reference GTF</td>
<td>–primary_reference_gtf</td>
<td>/projects/compsci/omics_share/mouse/GRCm39/transcriptome/annotation/ensembl/v105/Mus_musculus.GRCm39.105.gtf</td>
<td>Gene annotation for Ensembl 105 in GTF</td>
</tr></tbody></table><p>The default <code>nextflow</code> pipeline also include the DO
founder strains as a comma separated list:</p>
<table class="table"><colgroup><col width="58%"><col width="41%"></colgroup><thead><tr class="header"><th>Argument</th>
<th>Value</th>
</tr></thead><tbody><tr class="odd"><td>–strain</td>
<td>129S1_SvImJ,A_J,CAST_EiJ,NOD_ShiLtJ,NZO_HlLtJ,PWK_PhJ,WSB_EiJ</td>
</tr></tbody></table><p>To run the <code>nextflow</code> pipeline to generate a
pseudo-reference for DO mice on sumner2, use the following commands.</p>
<p>First, we load the <code>nextflow</code> module.</p>
<pre><code>module use --append /projects/omics_share/meta/modules
module load nextflow</code></pre>
<p>Next, change into a directory where you will be working. We suggest
using /flashscratch and creating a directory for this.</p>
<p>Create a variable for your user ID on sumner2:</p>
<pre><code>USERID=&lt;your User ID&gt;</code></pre>
<pre><code>cd /flashscratch/${USERID}</code></pre>
<p>Then use <code>git</code> to clone the cs-nf-pipelines
repository.</p>
<pre><code>git clone https://github.com/TheJacksonLaboratory/cs-nf-pipelines.git</code></pre>
<p>Next, we need to create a variable for the output directory name and
create the directory.</p>
<pre><code>OUTPUT_DIR=/flashscratch/${USERID}/gbrs
mkdir -p ${OUTPUT_DIR}</code></pre>
<p>The last step before running the pipeline is to create a temporary
working directory. This should be on /flashscratch since many large
files are created.</p>
<pre><code>WORKING_DIR=/flashscratch/${USERID}/tmp
mkdir -p ${WORKING_DIR}</code></pre>
<p>There are two types of arguments in the <code>nextflow</code> call
below. Arguments to <code>nextflow</code> have a single ‘-’. These
are:</p>
<table class="table"><colgroup><col width="32%"><col width="25%"><col width="41%"></colgroup><thead><tr class="header"><th>Argument</th>
<th>Value</th>
<th>Description</th>
</tr></thead><tbody><tr class="odd"><td>-profile</td>
<td>sumner2</td>
<td>This sets the resource values for the JAX sumner2 computing
cluster</td>
</tr><tr class="even"><td>-w</td>
<td>${WORKING_DIR}</td>
<td>The working directory for temporary files</td>
</tr></tbody></table><p>Arguments which start with ‘–’ go to the <code>nextflow</code>
pipeline. These are:</p>
<table class="table"><colgroup><col width="36%"><col width="24%"><col width="39%"></colgroup><thead><tr class="header"><th>Argument</th>
<th>Value</th>
<th>Description</th>
</tr></thead><tbody><tr class="odd"><td>–workflow</td>
<td>generate_pseudoreference</td>
<td>This tells <code>nextflow</code> which workflow to run</td>
</tr><tr class="even"><td>–pubdir</td>
<td>${OUTPUT_DIR}</td>
<td>The directory where output should be written</td>
</tr></tbody></table><p>With this preparation, we can run the <code>nextflow</code>
pipeline.</p>
<pre><code>nextflow cs-nf-pipelines/main.nf \
         -profile sumner2 \
         -w ${WORKING_DIR} \
         --workflow generate_pseudoreference \
         --pubdir ${OUTPUT_DIR}</code></pre>
<p>The entire process should take around 30 minutes. It should produce
the following output on the terminal:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>N E X T F L O W  ~  version 22.04.3
Launching `cs-nf-pipelines/main.nf` [gloomy_panini] DSL2 - revision: 6b82c2c015


 _____ _______ _     _     _______ _______     __   _  ______ _______      _____   _____  _______
   |   |_____|  \___/  ___ |       |______ ___ | \  | |  ____ |______ ___ |     | |_____| |______
 __|   |     | _/   \_     |_____  ______|     |  \_| |_____| ______|     |_____| |       ______|


GENERATE PSEUDOREFERENCE PARAMETER LOG

--comment:

Results Published to: /flashscratch/dgatti/gbrs
______________________________________________________
--workflow                      generate_pseudoreference
-w                              /flashscratch/dgatti/tmp
-c                              config/generate_pseudoreference.config
--snp_vcf                       /projects/compsci/omics_share/mouse/GRCm39/genome/annotation/snps_indels/rel_2112_v8/mgp_REL2021_snps.vcf.gz
--indel_vcf                     /projects/compsci/omics_share/mouse/GRCm39/genome/annotation/snps_indels/rel_2112_v8/mgp_REL2021_indels.vcf.gz
--primary_reference_fasta       /projects/compsci/omics_share/mouse/GRCm39/genome/sequence/ensembl/v105/Mus_musculus.GRCm39.dna.primary_assembly.fa
--primary_reference_gtf         /projects/compsci/omics_share/mouse/GRCm39/transcriptome/annotation/ensembl/v105/Mus_musculus.GRCm39.105.gtf
--gtf_biotype_include           protein_coding,lncRNA,IG_C_gene,IG_D_gene,IG_J_gene,IG_LV_gene,IG_V_gene,TR_C_gene,TR_D_gene,TR_J_gene,TR_V_gene
--strain                        129S1_SvImJ,A_J,CAST_EiJ,NOD_ShiLtJ,NZO_HlLtJ,PWK_PhJ,WSB_EiJ
--genome_version                39
--append_chromosomes            true
--diploid                       false
--keep_fails                    false
--pass_only                     false
--quality_filter
--region
--bed

--keep_intermediate             false

Project Directory: /gpfs/ctgs0/flashscratch/dgatti/nextflow/cs-nf-pipelines

Command line call:
nextflow cs-nf-pipelines/main.nf -profile sumner2 -w /flashscratch/dgatti/tmp --workflow generate_pseudoreference --pubdir /flashscratch/dgatti/gbrs
______________________________________________________

executor &gt;  slurm (72)
[9b/2185dd] process &gt; GENERATE_PSEUDOREFERENCE:FILTER_GTF (Filtering GTF)    [100%] 1 of 1 ✔
[86/d616d1] process &gt; GENERATE_PSEUDOREFERENCE:G2GTOOLS_VCF2VCI (WSB_EiJ)    [100%] 7 of 7 ✔
[4d/bce7df] process &gt; GENERATE_PSEUDOREFERENCE:G2GTOOLS_PATCH (WSB_EiJ)      [100%] 7 of 7 ✔
[5c/6c9c51] process &gt; GENERATE_PSEUDOREFERENCE:G2GTOOLS_TRANSFORM (WSB_EiJ)  [100%] 7 of 7 ✔
[97/91a924] process &gt; GENERATE_PSEUDOREFERENCE:SAMTOOLS_FAIDX_G2GTOOLS (W... [100%] 7 of 7 ✔
[22/380cb8] process &gt; GENERATE_PSEUDOREFERENCE:G2GTOOLS_CONVERT (CAST_EiJ)   [100%] 7 of 7 ✔
[6f/5a91f3] process &gt; GENERATE_PSEUDOREFERENCE:APPEND_DROPPED_CHROMS (CAS... [100%] 7 of 7 ✔
[fa/69434d] process &gt; GENERATE_PSEUDOREFERENCE:G2GTOOLS_GTF2DB (CAST_EiJ)    [100%] 7 of 7 ✔
[42/d32cda] process &gt; GENERATE_PSEUDOREFERENCE:G2GTOOLS_EXTRACT_GENES (CA... [100%] 7 of 7 ✔
[c5/15243b] process &gt; GENERATE_PSEUDOREFERENCE:G2GTOOLS_EXTRACT_TRANSCRIP... [100%] 7 of 7 ✔
[67/e947ad] process &gt; GENERATE_PSEUDOREFERENCE:G2GTOOLS_EXTRACT_EXONS (CA... [100%] 7 of 7 ✔
[54/203383] process &gt; GENERATE_PSEUDOREFERENCE:SAMTOOLS_FAIDX (primary_st... [100%] 1 of 1 ✔
Completed at: 14-Sep-2023 14:18:58
Duration    : 23m 40s
CPU hours   : 14.7
Succeeded   : 72</code></pre>
</div>
<p>Let’s look at the files in the output directory.</p>
<pre><code>ls ${OUTPUT_DIR}</code></pre>
<p>You should see the following files and directories:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>g2gtools  generate_pseudoreference_report.html  Mus_musculus.GRCm39.105.filtered.gtf  trace</code></pre>
</div>
<p>Look for the reference files in the ‘g2gtools’ directory.</p>
<pre><code>ls ${OUTPUT_DIR}/g2gtools</code></pre>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>129S1_SvImJ.39_DroppedChromAppended.gtf          NOD_ShiLtJ.39.genes.fa
129S1_SvImJ.39.exons.fa                          NOD_ShiLtJ.39.gtf.db
129S1_SvImJ.39.fa                                NOD_ShiLtJ.39.transcripts.fa
129S1_SvImJ.39.fa.fai                            NOD_ShiLtJ.39.vci.gz
129S1_SvImJ.39.genes.fa                          NOD_ShiLtJ.39.vci.gz.tbi
129S1_SvImJ.39.gtf.db                            NZO_HlLtJ.39_DroppedChromAppended.gtf
129S1_SvImJ.39.transcripts.fa                    NZO_HlLtJ.39.exons.fa
129S1_SvImJ.39.vci.gz                            NZO_HlLtJ.39.fa
129S1_SvImJ.39.vci.gz.tbi                        NZO_HlLtJ.39.fa.fai
A_J.39_DroppedChromAppended.gtf                  NZO_HlLtJ.39.genes.fa
A_J.39.exons.fa                                  NZO_HlLtJ.39.gtf.db
A_J.39.fa                                        NZO_HlLtJ.39.transcripts.fa
A_J.39.fa.fai                                    NZO_HlLtJ.39.vci.gz
A_J.39.genes.fa                                  NZO_HlLtJ.39.vci.gz.tbi
A_J.39.gtf.db                                    PWK_PhJ.39_DroppedChromAppended.gtf
A_J.39.transcripts.fa                            PWK_PhJ.39.exons.fa
A_J.39.vci.gz                                    PWK_PhJ.39.fa
A_J.39.vci.gz.tbi                                PWK_PhJ.39.fa.fai
CAST_EiJ.39_DroppedChromAppended.gtf             PWK_PhJ.39.genes.fa
CAST_EiJ.39.exons.fa                             PWK_PhJ.39.gtf.db
CAST_EiJ.39.fa                                   PWK_PhJ.39.transcripts.fa
CAST_EiJ.39.fa.fai                               PWK_PhJ.39.vci.gz
CAST_EiJ.39.genes.fa                             PWK_PhJ.39.vci.gz.tbi
CAST_EiJ.39.gtf.db                               WSB_EiJ.39_DroppedChromAppended.gtf
CAST_EiJ.39.transcripts.fa                       WSB_EiJ.39.exons.fa
CAST_EiJ.39.vci.gz                               WSB_EiJ.39.fa
CAST_EiJ.39.vci.gz.tbi                           WSB_EiJ.39.fa.fai
Mus_musculus.GRCm39.dna.primary_assembly.fa.fai  WSB_EiJ.39.genes.fa
NOD_ShiLtJ.39_DroppedChromAppended.gtf           WSB_EiJ.39.gtf.db
NOD_ShiLtJ.39.exons.fa                           WSB_EiJ.39.transcripts.fa
NOD_ShiLtJ.39.fa                                 WSB_EiJ.39.vci.gz
NOD_ShiLtJ.39.fa.fai                             WSB_EiJ.39.vci.gz.tbi</code></pre>
</div>
</section><section id="creating-pseudo-reference-genomes-and-transcriptomes-for-other-strains"><h2 class="section-heading">Creating Pseudo-Reference Genomes and Transcriptomes for Other
Strains<a class="anchor" aria-label="anchor" href="#creating-pseudo-reference-genomes-and-transcriptomes-for-other-strains"></a>
</h2>
<hr class="half-width"><p>If you are working with a cross comprised of other strains, you can
change the ’–strain</p>
<div id="challenge-1-viewing-strain-names-in-a-vcf" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1-viewing-strain-names-in-a-vcf" class="callout-inner">
<h3 class="callout-title">Challenge 1: Viewing Strain Names in a
VCF<a class="anchor" aria-label="anchor" href="#challenge-1-viewing-strain-names-in-a-vcf"></a>
</h3>
<div class="callout-content">
<p>There are 50 strains in the VCF SNP file. What are their names? The
strain names are stored in the VCF header. The <code>samtools</code>
container also includes a tool called <code>tabix</code>, which queries
and indexes tab-delimited files. Run the <code>samtools</code> container
and look at the tabix help page. Find a command that will print out the
SNP VCF header and find the strain names.</p>
<pre><code>singularity run ${SAMTOOLS} tabix --help</code></pre>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">Output</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>Call <code>tabix</code> with the ‘–only-header’ option and pass in
the SNP VCF. In this course, we are using long option names. You could
also use ‘-H’ to achieve the same result.</p>
<pre><code>$ singularity run ${SAMTOOLS} tabix --only-header ${VCF_SNPS}</code></pre>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="co">##fileformat=VCFv4.2</span></span>
<span><span class="co">##FILTER=&lt;ID=PASS,Description="All filters passed"&gt;</span></span>
<span><span class="co">##bcftoolsVersion=1.13+htslib-1.13</span></span>
<span><span class="co">##bcftoolsCommand=mpileup -f Mus_musculus.GRCm39.dna.toplevel.fa.gz -b samples -g 10 -a FORMAT/DP,FORMAT/AD,FORMAT/ADF,FORMAT/ADR,FORMAT/SP,INFO/AD -E -Q 0 -pm 3 -F 0.25 -d 500</span></span>
<span><span class="co">##reference=Mus_musculus.GRCm39.dna.toplevel.fa.gz</span></span>
<span><span class="co">##contig=&lt;ID=1,length=195154279&gt;</span></span>
<span><span class="co">##contig=&lt;ID=2,length=181755017&gt;</span></span>
<span><span class="co">##contig=&lt;ID=3,length=159745316&gt;</span></span>
<span><span class="co">##contig=&lt;ID=4,length=156860686&gt;</span></span>
<span><span class="co">##contig=&lt;ID=5,length=151758149&gt;</span></span>
<span><span class="co">##contig=&lt;ID=6,length=149588044&gt;</span></span>
<span><span class="co">##contig=&lt;ID=7,length=144995196&gt;</span></span>
<span><span class="co">##contig=&lt;ID=8,length=130127694&gt;</span></span>
<span><span class="co">##contig=&lt;ID=9,length=124359700&gt;</span></span>
<span><span class="co">##contig=&lt;ID=10,length=130530862&gt;</span></span>
<span><span class="co">##contig=&lt;ID=11,length=121973369&gt;</span></span>
<span><span class="co">##contig=&lt;ID=12,length=120092757&gt;</span></span>
<span><span class="co">##contig=&lt;ID=13,length=120883175&gt;</span></span>
<span><span class="co">##contig=&lt;ID=14,length=125139656&gt;</span></span>
<span><span class="co">##contig=&lt;ID=15,length=104073951&gt;</span></span>
<span><span class="co">##contig=&lt;ID=16,length=98008968&gt;</span></span>
<span><span class="co">##contig=&lt;ID=17,length=95294699&gt;</span></span>
<span><span class="co">##contig=&lt;ID=18,length=90720763&gt;</span></span>
<span><span class="co">##contig=&lt;ID=19,length=61420004&gt;</span></span>
<span><span class="co">##contig=&lt;ID=X,length=169476592&gt;</span></span>
<span><span class="co">##ALT=&lt;ID=*,Description="Represents allele(s) other than observed."&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=INDEL,Number=0,Type=Flag,Description="Indicates that the variant is an INDEL."&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=IDV,Number=1,Type=Integer,Description="Maximum number of raw reads supporting an indel"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=IMF,Number=1,Type=Float,Description="Maximum fraction of raw reads supporting an indel"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=DP,Number=1,Type=Integer,Description="Raw read depth"&gt;</span></span>
<span><span class="co">##FORMAT=&lt;ID=PL,Number=G,Type=Integer,Description="List of Phred-scaled genotype likelihoods"&gt;</span></span>
<span><span class="co">##FORMAT=&lt;ID=DP,Number=1,Type=Integer,Description="Number of high-quality bases"&gt;</span></span>
<span><span class="co">##FORMAT=&lt;ID=AD,Number=R,Type=Integer,Description="Allelic depths (high-quality bases)"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=AD,Number=R,Type=Integer,Description="Total allelic depths (high-quality bases)"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=END,Number=1,Type=Integer,Description="End position of the variant described in this record"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=MinDP,Number=1,Type=Integer,Description="Minimum per-sample depth in this gVCF block"&gt;</span></span>
<span><span class="co">##FORMAT=&lt;ID=GT,Number=1,Type=String,Description="Genotype"&gt;</span></span>
<span><span class="co">##FORMAT=&lt;ID=GQ,Number=1,Type=Integer,Description="Phred-scaled Genotype Quality"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=DP4,Number=4,Type=Integer,Description="Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=MQ,Number=1,Type=Integer,Description="Average mapping quality"&gt;</span></span>
<span><span class="co">##bcftools_callCommand=call -mAv -f GQ,GP -p 0.99; Date=Wed Aug 11 21:20:03 2021</span></span>
<span><span class="co">##bcftools_normCommand=norm --fasta-ref Mus_musculus.GRCm39.dna.toplevel.fa.gz -m +indels; Date=Fri Aug 13 11:11:49 2021</span></span>
<span><span class="co">##FORMAT=&lt;ID=FI,Number=1,Type=Integer,Description="High confidence (1) or low confidence (0) based on soft filtering values"&gt;</span></span>
<span><span class="co">##FILTER=&lt;ID=LowQual,Description="Low quality variants"&gt;</span></span>
<span><span class="co">##VEP="v104" time="2021-08-30 23:27:00" cache="mus_musculus/104_GRCm39" ensembl-funcgen=104.59ae779 ensembl-variation=104.6154f8b ensembl=104.1af1dce ensembl-io=104.1d3bb6e assembly="GRCm39" dbSNP="150" gencode="GENCODE M27" regbuild="1.0" sift="sift"</span></span>
<span><span class="co">##INFO=&lt;ID=CSQ,Number=.,Type=String,Description="Consequence annotations from Ensembl VEP. Format: Allele|Consequence|IMPACT|SYMBOL|Gene|Feature_type|Feature|BIOTYPE|EXON|INTRON|HGVSc|HGVSp|cDNA_position|CDS_position|Protein_position|Amino_acids|Codons|Existing_variation|DISTANCE|STRAND|FLAGS|VARIANT_CLASS|SYMBOL_SOURCE|HGNC_ID|SIFT|MOTIF_NAME|MOTIF_POS|HIGH_INF_POS|MOTIF_SCORE_CHANGE|TRANSCRIPTION_FACTORS"&gt;</span></span>
<span><span class="co">##bcftools_viewVersion=1.13+htslib-1.13</span></span>
<span><span class="co">##bcftools_viewCommand=view -i 'FORMAT/FI[*] = 1' mgp_REL2021_snps.vcf.gz; Date=Sat Dec 18 19:08:09 2021</span></span>
<span><span class="co">##bcftools_annotateVersion=1.13+htslib-1.13</span></span>
<span><span class="co">##bcftools_annotateCommand=annotate -x INFO/VDB,INFO/SGB,INFO/RPBZ,INFO/MQBZ,INFO/MQBZ,INFO/MQSBZ,INFO/BQBZ,INFO/SCBZ,INFO/FS,INFO/MQOF,INFO/AC,INFO/AN,FORMAT/SP,FORMAT/ADF,FORMAT/ADR,FORMAT/GP; Date=Sat Dec 18 19:08:09 2021</span></span>
<span><span class="co">##INFO=&lt;ID=AC,Number=A,Type=Integer,Description="Allele count in genotypes"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=AN,Number=1,Type=Integer,Description="Total number of alleles in called genotypes"&gt;</span></span>
<span><span class="co">##bcftools_viewCommand=view -a -Oz -o final_mgp_REL2021_snps.vcf.gz; Date=Sat Dec 18 19:08:09 2021</span></span>
<span><span class="co">##bcftools_annotateCommand=annotate -x INFO/MQ0F -Oz -o final_mgp_REL2021_snps.vcf.gz mgp_REL2021_snps.vcf.gz; Date=Mon Dec 20 07:12:23 2021</span></span>
<span><span class="co">#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  129P2_OlaHsd    129S1_SvImJ    129S5SvEvBrd     A_J     AKR_J   B10.RIII        BALB_cByJ       BALB_cJ BTBR_T+_Itpr3tf_J       BUB_BnJC3H_HeH  C3H_HeJ C57BL_10J       C57BL_10SnJ     C57BL_6NJ       C57BR_cdJ       C57L_J  C58_J   CAST_EiJCBA_J   CE_J    CZECHII_EiJ     DBA_1J  DBA_2J  FVB_NJ  I_LnJ   JF1_MsJ KK_HiJ  LEWES_EiJ       LG_J   LP_J     MAMy_J  MOLF_EiJ        NOD_ShiLtJ      NON_LtJ NZB_B1NJ        NZO_HlLtJ       NZW_LacJ       PL_J     PWK_PhJ QSi3    QSi5    RF_J    RIIIS_J SEA_GnJ SJL_J   SM_J    SPRET_EiJ       ST_bJ   SWR_J  WSB_EiJ  ZALENDE_EiJ</span></span></code></pre>
</div>
<p>The VCF header contains information about the file format version,
chromosomes, metadata, and tools used to create the VCF. The column
names are on the last line of the VCF header and this line contains the
strain names. If you had a cross comprised of other strains, you would
query the VCF to find their names.</p>
</div>
</div>
</div>
</div>
</section><section id="step-by-step-pseudo-reference-generation"><h2 class="section-heading">Step-by-Step Pseudo-Reference Generation<a class="anchor" aria-label="anchor" href="#step-by-step-pseudo-reference-generation"></a>
</h2>
<hr class="half-width"><p>In general, most users will not use this strategy. We have included
this section to document the steps for advanced users who may want to
modify specific steps in the pipeline.</p>
<p>The first step in creating the reference files for GBRS is to create
a set of genomes and transcriptomes with SNPs and Indels from other
strains inserted into the reference genome.</p>
<div class="section level3">
<h3 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a></h3>
<p>We will use the g2gtools container stored in the public reference
area on sumner2.</p>
<pre><code>G2GTOOLS=/projects/omics_share/meta/containers/quay.io-jaxcompsci-g2gtools-74926ad.img</code></pre>
<p>We will also use a tool called <a href="http://www.htslib.org/doc/samtools.html" class="external-link">samtools</a> to query and
modify sequence and annotation files.</p>
<pre><code>SAMTOOLS=/projects/omics_share/meta/containers/quay.io-biocontainers-samtools-1.14--hb421002_0.img</code></pre>
<p>We need two input files and we will produce one output file per
strain.</p>
<p>The first file is the reference genome in FASTA format.</p>
<pre><code>REF_FASTA=/projects/omics_share/mouse/GRCm39/genome/sequence/ensembl/v105/Mus_musculus.GRCm39.dna.primary_assembly.fa</code></pre>
<p>The second file is a VCF containing the SNPs of one or more strains.
This file should contain the strain(s) that you have in your mouse
cross.</p>
<pre><code>VCF_INDELS=/projects/omics_share/mouse/GRCm39/genome/annotation/snps_indels/rel_2112_v8/mgp_REL2021_indels.vcf.gz
VCF_SNPS=/projects/omics_share/mouse/GRCm39/genome/annotation/snps_indels/rel_2112_v8/mgp_REL2021_snps.vcf.gz</code></pre>
<p>Next, we need to pass in the name of the strain for which we are
creating the VCI file. In this case, we will use DBA/2J.</p>
<pre><code><span><span class="va">STRAIN</span><span class="op">=</span><span class="va">DBA_2J</span></span></code></pre>
<div id="accordionInstructor1" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor1" aria-expanded="false" aria-controls="collapseInstructor1">
  <h3 class="accordion-header" id="headingInstructor1">
<div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div>Instructor Note</h3>
</button>
<div id="collapseInstructor1" class="accordion-collapse collapse" aria-labelledby="headingInstructor1" data-bs-parent="#accordionInstructor1">
<div class="accordion-body">
<p>This Challenge is used in above as well in “Creating Pseudo-Reference
Genomes and Transcriptomes for Other Strains”.</p>
</div>
</div>
</div>
</div>
<div id="challenge-2-viewing-strain-names-in-a-vcf" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2-viewing-strain-names-in-a-vcf" class="callout-inner">
<h3 class="callout-title">Challenge 2: Viewing Strain Names in a
VCF<a class="anchor" aria-label="anchor" href="#challenge-2-viewing-strain-names-in-a-vcf"></a>
</h3>
<div class="callout-content">
<p>There are 50 strains in the VCF SNP file. What are their names? The
strain names are stored in the VCF header. The <code>samtools</code>
container also includes a tool called <code>tabix</code>, which queries
and indexes tab-delimited files. Run the <code>samtools</code> container
and look at the tabix help page. Find a command that will print out the
SNP VCF header and find the strain names.</p>
<pre><code>singularity run ${SAMTOOLS} tabix --help</code></pre>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">Output</h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>Call <code>tabix</code> with the ‘–only-header’ option and pass in
the SNP VCF. In this course, we are using long option names. You could
also use ‘-H’ to achieve the same result.</p>
<pre><code>$ singularity run ${SAMTOOLS} tabix --only-header ${VCF_SNPS}</code></pre>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="co">##fileformat=VCFv4.2</span></span>
<span><span class="co">##FILTER=&lt;ID=PASS,Description="All filters passed"&gt;</span></span>
<span><span class="co">##bcftoolsVersion=1.13+htslib-1.13</span></span>
<span><span class="co">##bcftoolsCommand=mpileup -f Mus_musculus.GRCm39.dna.toplevel.fa.gz -b samples -g 10 -a FORMAT/DP,FORMAT/AD,FORMAT/ADF,FORMAT/ADR,FORMAT/SP,INFO/AD -E -Q 0 -pm 3 -F 0.25 -d 500</span></span>
<span><span class="co">##reference=Mus_musculus.GRCm39.dna.toplevel.fa.gz</span></span>
<span><span class="co">##contig=&lt;ID=1,length=195154279&gt;</span></span>
<span><span class="co">##contig=&lt;ID=2,length=181755017&gt;</span></span>
<span><span class="co">##contig=&lt;ID=3,length=159745316&gt;</span></span>
<span><span class="co">##contig=&lt;ID=4,length=156860686&gt;</span></span>
<span><span class="co">##contig=&lt;ID=5,length=151758149&gt;</span></span>
<span><span class="co">##contig=&lt;ID=6,length=149588044&gt;</span></span>
<span><span class="co">##contig=&lt;ID=7,length=144995196&gt;</span></span>
<span><span class="co">##contig=&lt;ID=8,length=130127694&gt;</span></span>
<span><span class="co">##contig=&lt;ID=9,length=124359700&gt;</span></span>
<span><span class="co">##contig=&lt;ID=10,length=130530862&gt;</span></span>
<span><span class="co">##contig=&lt;ID=11,length=121973369&gt;</span></span>
<span><span class="co">##contig=&lt;ID=12,length=120092757&gt;</span></span>
<span><span class="co">##contig=&lt;ID=13,length=120883175&gt;</span></span>
<span><span class="co">##contig=&lt;ID=14,length=125139656&gt;</span></span>
<span><span class="co">##contig=&lt;ID=15,length=104073951&gt;</span></span>
<span><span class="co">##contig=&lt;ID=16,length=98008968&gt;</span></span>
<span><span class="co">##contig=&lt;ID=17,length=95294699&gt;</span></span>
<span><span class="co">##contig=&lt;ID=18,length=90720763&gt;</span></span>
<span><span class="co">##contig=&lt;ID=19,length=61420004&gt;</span></span>
<span><span class="co">##contig=&lt;ID=X,length=169476592&gt;</span></span>
<span><span class="co">##ALT=&lt;ID=*,Description="Represents allele(s) other than observed."&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=INDEL,Number=0,Type=Flag,Description="Indicates that the variant is an INDEL."&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=IDV,Number=1,Type=Integer,Description="Maximum number of raw reads supporting an indel"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=IMF,Number=1,Type=Float,Description="Maximum fraction of raw reads supporting an indel"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=DP,Number=1,Type=Integer,Description="Raw read depth"&gt;</span></span>
<span><span class="co">##FORMAT=&lt;ID=PL,Number=G,Type=Integer,Description="List of Phred-scaled genotype likelihoods"&gt;</span></span>
<span><span class="co">##FORMAT=&lt;ID=DP,Number=1,Type=Integer,Description="Number of high-quality bases"&gt;</span></span>
<span><span class="co">##FORMAT=&lt;ID=AD,Number=R,Type=Integer,Description="Allelic depths (high-quality bases)"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=AD,Number=R,Type=Integer,Description="Total allelic depths (high-quality bases)"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=END,Number=1,Type=Integer,Description="End position of the variant described in this record"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=MinDP,Number=1,Type=Integer,Description="Minimum per-sample depth in this gVCF block"&gt;</span></span>
<span><span class="co">##FORMAT=&lt;ID=GT,Number=1,Type=String,Description="Genotype"&gt;</span></span>
<span><span class="co">##FORMAT=&lt;ID=GQ,Number=1,Type=Integer,Description="Phred-scaled Genotype Quality"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=DP4,Number=4,Type=Integer,Description="Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=MQ,Number=1,Type=Integer,Description="Average mapping quality"&gt;</span></span>
<span><span class="co">##bcftools_callCommand=call -mAv -f GQ,GP -p 0.99; Date=Wed Aug 11 21:20:03 2021</span></span>
<span><span class="co">##bcftools_normCommand=norm --fasta-ref Mus_musculus.GRCm39.dna.toplevel.fa.gz -m +indels; Date=Fri Aug 13 11:11:49 2021</span></span>
<span><span class="co">##FORMAT=&lt;ID=FI,Number=1,Type=Integer,Description="High confidence (1) or low confidence (0) based on soft filtering values"&gt;</span></span>
<span><span class="co">##FILTER=&lt;ID=LowQual,Description="Low quality variants"&gt;</span></span>
<span><span class="co">##VEP="v104" time="2021-08-30 23:27:00" cache="mus_musculus/104_GRCm39" ensembl-funcgen=104.59ae779 ensembl-variation=104.6154f8b ensembl=104.1af1dce ensembl-io=104.1d3bb6e assembly="GRCm39" dbSNP="150" gencode="GENCODE M27" regbuild="1.0" sift="sift"</span></span>
<span><span class="co">##INFO=&lt;ID=CSQ,Number=.,Type=String,Description="Consequence annotations from Ensembl VEP. Format: Allele|Consequence|IMPACT|SYMBOL|Gene|Feature_type|Feature|BIOTYPE|EXON|INTRON|HGVSc|HGVSp|cDNA_position|CDS_position|Protein_position|Amino_acids|Codons|Existing_variation|DISTANCE|STRAND|FLAGS|VARIANT_CLASS|SYMBOL_SOURCE|HGNC_ID|SIFT|MOTIF_NAME|MOTIF_POS|HIGH_INF_POS|MOTIF_SCORE_CHANGE|TRANSCRIPTION_FACTORS"&gt;</span></span>
<span><span class="co">##bcftools_viewVersion=1.13+htslib-1.13</span></span>
<span><span class="co">##bcftools_viewCommand=view -i 'FORMAT/FI[*] = 1' mgp_REL2021_snps.vcf.gz; Date=Sat Dec 18 19:08:09 2021</span></span>
<span><span class="co">##bcftools_annotateVersion=1.13+htslib-1.13</span></span>
<span><span class="co">##bcftools_annotateCommand=annotate -x INFO/VDB,INFO/SGB,INFO/RPBZ,INFO/MQBZ,INFO/MQBZ,INFO/MQSBZ,INFO/BQBZ,INFO/SCBZ,INFO/FS,INFO/MQOF,INFO/AC,INFO/AN,FORMAT/SP,FORMAT/ADF,FORMAT/ADR,FORMAT/GP; Date=Sat Dec 18 19:08:09 2021</span></span>
<span><span class="co">##INFO=&lt;ID=AC,Number=A,Type=Integer,Description="Allele count in genotypes"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=AN,Number=1,Type=Integer,Description="Total number of alleles in called genotypes"&gt;</span></span>
<span><span class="co">##bcftools_viewCommand=view -a -Oz -o final_mgp_REL2021_snps.vcf.gz; Date=Sat Dec 18 19:08:09 2021</span></span>
<span><span class="co">##bcftools_annotateCommand=annotate -x INFO/MQ0F -Oz -o final_mgp_REL2021_snps.vcf.gz mgp_REL2021_snps.vcf.gz; Date=Mon Dec 20 07:12:23 2021</span></span>
<span><span class="co">#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  129P2_OlaHsd    129S1_SvImJ    129S5SvEvBrd     A_J     AKR_J   B10.RIII        BALB_cByJ       BALB_cJ BTBR_T+_Itpr3tf_J       BUB_BnJC3H_HeH  C3H_HeJ C57BL_10J       C57BL_10SnJ     C57BL_6NJ       C57BR_cdJ       C57L_J  C58_J   CAST_EiJCBA_J   CE_J    CZECHII_EiJ     DBA_1J  DBA_2J  FVB_NJ  I_LnJ   JF1_MsJ KK_HiJ  LEWES_EiJ       LG_J   LP_J     MAMy_J  MOLF_EiJ        NOD_ShiLtJ      NON_LtJ NZB_B1NJ        NZO_HlLtJ       NZW_LacJ       PL_J     PWK_PhJ QSi3    QSi5    RF_J    RIIIS_J SEA_GnJ SJL_J   SM_J    SPRET_EiJ       ST_bJ   SWR_J  WSB_EiJ  ZALENDE_EiJ</span></span>
<span></span></code></pre>
</div>
<p>The VCF header contains information about the file format version,
chromosomes, metadata, and tools used to create the VCF. The column
names are on the last line of the VCF header and this line contains the
strain names. If you had a cross comprised of other strains, you would
query the VCF to find their names.</p>
</div>
</div>
</div>
</div>
<p>We will work on the sumner2 /flashscratch area. Create a directory
with your user ID. Then create a directory called ‘gbrs’. For
example:</p>
<pre><code>mkdir -p /flashscratch/dgatti/gbrs/${STRAIN}</code></pre>
<p>Then change into the directory that you just created.</p>
<pre><code><span><span class="va">cd</span> <span class="op">/</span><span class="va">flashscratch</span><span class="op">/</span><span class="va">dgatti</span><span class="op">/</span><span class="va">gbrs</span></span></code></pre>
<p>Load in the Singularity software.</p>
<pre><code>module load singularity</code></pre>
</div>
<div class="section level3">
<h3 id="use-g2gtools-vcf2vci-to-gather-strain-specific-snps-and-indels">Use <code>g2gtools vcf2vci</code> to Gather Strain-Specific SNPs and
Indels<a class="anchor" aria-label="anchor" href="#use-g2gtools-vcf2vci-to-gather-strain-specific-snps-and-indels"></a></h3>
<p>In the first step, we insert indels from a specific strain into the
reference genome using the <code>vcf2vci</code> command. This is written
out to a text file in a format called “VCI”.</p>
<p>The arguments are:</p>
<pre><code>$ g2gtools vcf2vci --help
╭─ Options ───────────────────────────────────────────────────────────────────────────────────────────────╮
│ *  --vcf       -i      FILE     VCF files can seperate files by "," or have multiple -i [default: None]  │
│                                 [required]                                                               │
│ *  --fasta     -f      FILE     Fasta file matching VCF information [default: None] [required]           │
│ *  --strain    -s      TEXT     Name of strain/sample (column in VCF file) [default: None] [required]    │
│    --output    -o      FILE     Name of output file [default: None]                                      │
│    --diploid   -d               Create diploid VCI file                                                  │
│    --keep                       Keep track of VCF lines that could not be converted to VCI file          │
│    --pass                       Use only VCF lines that have a PASS for the filter value                 │
│    --quality                    Filter on quality, FI=PASS                                               │
│    --no-bgzip                   DO NOT compress and index output                                         │
│    --verbose   -v      INTEGER  specify multiple times for more verbose output [default: 0]              │
│    --help                       Show this message and exit.                                              │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────╯</code></pre>
<p>We will use a subset of the arguments, passing in the reference FASTA
file, the SNPD and indel VCFs, the strain name to search for in the
indel VCF, and the output file path.</p>
<p><strong>Inputs</strong> | Argument | Description | |———-|————-| |
–fasta | path to the GRCm39 reference FASTA file | | –vcf | path(s) to
the Sanger VCFs for SNPs and indels. Both can be passed in. |</p>
<p><strong>Output</strong> | Argument | Description | |———-|————-| |
–output | A VCI file, which is a custom file format, akin to a BED
format, which lists the positions and sequences of indels. Note that,
while we list the output file extension as “vci”, <code>g2gtools</code>
will gzip and index the file, so we will need to add the ‘.gz’ extension
when we use the file in downstream commands. |</p>
<pre><code>STRAIN_VCI=${STRAIN}/REF-to-${STRAIN}.vci
singularity run ${G2GTOOLS} g2gtools vcf2vci \
                            --fasta ${REF_FASTA} \
                            --vcf ${VCF_INDELS} \
                            --vcf ${VCF_SNPS} \
                            --strain ${STRAIN}  \
                            --output ${STRAIN_VCI}</code></pre>
</div>
<div class="section level3">
<h3 id="use-g2gtools-patch-to-insert-strain-specific-snps-into-reference">Use <code>g2gtools patch</code> to Insert Strain-Specific SNPs into
Reference<a class="anchor" aria-label="anchor" href="#use-g2gtools-patch-to-insert-strain-specific-snps-into-reference"></a></h3>
<p>Next, we insert SNPs from a specific strain into the reference genome
using the ‘patch’ command.</p>
<p>The arguments are:</p>
<pre><code>$ g2gtools patch --help
╭─ Options ───────────────────────────────────────────────────────────────────────────────────────╮
│ *  --input    -i      FILE     Fasta file to extract from [default: None] [required]             │
│ *  --vci      -c      FILE     VCI File to use [default: None] [required]                        │
│    --output   -o      FILE     Name of output file [default: None]                               │
│    --bed      -b      FILE     BED file name [default: None]                                     │
│    --region   -r      TEXT     Region to extract in chromosome:start-end format [default: None]  │
│    --reverse                   Reverse the direction of VCI file                                 │
│    --bgzip                     compress and index output                                         │
│    --verbose  -v      INTEGER  specify multiple times for more verbose output [default: 0]       │
│    --help                      Show this message and exit.                                       │
╰─────────────────────────────────────────────────────────────────────────────────────────────────╯</code></pre>
<p>We will use the following arguments:</p>
<p><strong>Inputs</strong> | Argument | Description | |———-|————-| |
–input | path to the GRCm39 reference FASTA file | | –vci | path to the
<strong>gzipped</strong> VCI file created by <code>vcf2vci</code>. |</p>
<p><strong>Output</strong> | Argument | Description | |———-|————-| |
–output | path to patched FASTA file with SNPs inserted into the
reference genome sequence. |</p>
<pre><code>PATCHED_FASTA=${STRAIN}/${STRAIN}.patched.fa
singularity run ${G2GTOOLS} g2gtools patch \
                            --input ${REF_FASTA} \
                            --vci ${STRAIN_VCI}.gz \
                            --output ${PATCHED_FASTA}</code></pre>
</div>
<div class="section level3">
<h3 id="use-g2gtools-transform-to-insert-strain-specific-indels-into-strain-fasta">Use <code>g2gtools transform</code> to Insert Strain-Specific Indels
into Strain FASTA<a class="anchor" aria-label="anchor" href="#use-g2gtools-transform-to-insert-strain-specific-indels-into-strain-fasta"></a></h3>
<p>The ‘transform’ function takes the strain-specific SNP-patched FASTA
file, inserts indels from the VCI file, and outputs a FASTA file. This
FASTA file contains the inferred sequence of a specific strain.</p>
<p>Next, we insert SNPs from a specific strain into the reference genome
using the ‘patch’ command.</p>
<p>The arguments are:</p>
<pre><code>$ g2gtools transform --help
╭─ Options ───────────────────────────────────────────────────────────────────────────────────────╮
│ *  --input    -i      FILE     Fasta file to extract from [default: None] [required]             │
│ *  --vci      -c      FILE     VCI File to use [default: None] [required]                        │
│    --output   -o      FILE     Name of output file [default: None]                               │
│    --bed      -b      FILE     BED file name [default: None]                                     │
│    --region   -r      TEXT     Region to extract in chromosome:start-end format [default: None]  │
│    --reverse                   Reverse the direction of VCI file                                 │
│    --bgzip                     compress and index output                                         │
│    --verbose  -v      INTEGER  specify multiple times for more verbose output [default: 0]       │
│    --help                      Show this message and exit.                                       │
╰─────────────────────────────────────────────────────────────────────────────────────────────────╯</code></pre>
<p>We will use the following arguments:</p>
<p><strong>Inputs</strong> | Argument | Description | |———-|————-| |
–input | path to the strain-specific patched FASTA file created by
<code>patch</code>. | | –vci | path to the <strong>gzipped</strong> VCI
file created by <code>vcf2vci</code>. |</p>
<p><strong>Output</strong> | Argument | Description | |———-|————-| |
–output | path to transformed FASTA file with SNPs and indels inserted
into the reference genome sequence. |</p>
<p>We need to create a genome version variable to create informative
file names.</p>
<pre><code><span><span class="va">GENOME_VERSION</span><span class="op">=</span><span class="va">GRCm39</span></span></code></pre>
<p>Next, we will run the <code>transform</code> command.</p>
<pre><code>STRAIN_FASTA=${STRAIN}/${STRAIN}.${GENOME_VERSION}.fa
singularity run ${G2GTOOLS} g2gtools transform \
                            --input ${PATCHED_FASTA} \
                            --vci ${STRAIN_VCI}.gz \
                            --output ${STRAIN_FASTA}</code></pre>
<p>We now have a pseudo-reference genome in the strain-specific FASTA
file.</p>
</div>
<div class="section level3">
<h3 id="use-samtools-faidx-to-index-the-strain-specific-fasta-file">Use <code>samtools faidx</code> to Index the Strain-Specific FASTA
file<a class="anchor" aria-label="anchor" href="#use-samtools-faidx-to-index-the-strain-specific-fasta-file"></a></h3>
<p>This is a step which uses the <a href="https://www.htslib.org/doc/samtools.html" class="external-link">samtools</a> suite of
tools to index the FASTA file.</p>
<p>Remember that we created a variable for the <code>samtools</code>
Singularity container at the top of the lesson.</p>
<p>Then we will use <code>samtools</code> to index the FASTA file.</p>
<pre><code>singularity run ${SAMTOOLS} samtools faidx ${STRAIN_FASTA}</code></pre>
</div>
<div class="section level3">
<h3 id="use-g2gtools-convert-to-create-a-strain-specific-gtf-file">Use <code>g2gtools convert</code> to Create a Strain-Specific GTF
File<a class="anchor" aria-label="anchor" href="#use-g2gtools-convert-to-create-a-strain-specific-gtf-file"></a></h3>
<p>The <code>convert</code> function takes the strain-specific FASTA
file, the VCI file, and the reference GTF and creates a strain-specific
annotation file in GTF.</p>
<p>The arguments are:</p>
<pre><code>$ g2gtools convert --help
╭─ Options ────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ *  --input-file   -i      FILE                   Input file to convert to new coordinates [default: None] [required]  │
│ *  --vci-file     -c      FILE                   VCI file [default: None] [required]                                  │
│ *  --file-format  -f      [BAM|SAM|GFF|GTF|BED]  Input file format [default: None] [required]                         │
│    --output       -o      FILE                   Name of output file [default: None]                                  │
│    --reverse      -r                             Reverse the direction of the conversion                              │
│    --verbose      -v      INTEGER                specify multiple times for more verbose output [default: 0]          │
│    --help                                        Show this message and exit.                                          │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯</code></pre>
<p>We will use the following arguments:</p>
<p><strong>Inputs</strong> | Argument | Description | |———-|————-| |
–input-file | path to the reference GTF. | | –vci-file | path to the
<strong>gzipped</strong> VCI file created by <code>vcf2vci</code>. | |
–file-format | string indicating a GTF file. Lower case since it will be
used as a file name extension. |</p>
<p><strong>Output</strong> | Argument | Description | |———-|————-| |
–output | path to converted GTF file with SNPs and indels inserted into
the reference transcript and coordinates shifted to the strain-specific
coordinates. |</p>
<p>Next, we need the path to the reference GTF. We have selected the
Ensembl 105 annotation for this lesson.</p>
<pre><code>REF_GTF=/projects/omics_share/mouse/GRCm39/transcriptome/annotation/ensembl/v105/Mus_musculus.GRCm39.105.chr.gtf</code></pre>
<p>We also need to specify the annotation file format in lower case
because it will be used as a file name extension.</p>
<pre><code><span><span class="va">ANNOT_FORMAT</span><span class="op">=</span><span class="va">gtf</span></span></code></pre>
<pre><code>STRAIN_GTF=${STRAIN}/${STRAIN}.${GENOME_VERSION}.${ANNOT_FORMAT}
singularity run ${G2GTOOLS} g2gtools convert \
                            --input-file ${REF_GTF} \
                            --vci-file ${STRAIN_VCI}.gz \
                            --file-format ${ANNOT_FORMAT} \
                            --output ${STRAIN_GTF}</code></pre>
<p>We now have two key files that are used by GBRS:</p>
<ul><li>the strain-specific FASTA file, which contains the strain’s inferred
sequence, and</li>
<li>the strain-specific GTF file, which contains the strain’s inferred
annotation.</li>
</ul><blockquote>
<p>DMG: Stopped here.</p>
</blockquote>
<div id="challenge-3-create-a-set-of-reference-files-for-another-strain-in-the-vcf." class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-3-create-a-set-of-reference-files-for-another-strain-in-the-vcf." class="callout-inner">
<h3 class="callout-title">Challenge 3: Create a set of reference files
for another strain in the VCF.<a class="anchor" aria-label="anchor" href="#challenge-3-create-a-set-of-reference-files-for-another-strain-in-the-vcf."></a>
</h3>
<div class="callout-content">
<p>You will need to look back at Challenge 1 to see how to list the
strains in the VCFs. You can query either the SNP or Indel VCF. Create
an entire script on your compute instance to run the commands above</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">Output</h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "This new lesson looks good"</code></pre>
</div>
</div>
</div>
</div>
</div>
<pre><code>g2gtools transform -i ${STRAIN}/${STRAIN}.patched.fa -c ${STRAIN}/REF-to-${STRAIN}.chain -o ${STRAIN}/${STRAIN}.fa
g2gtools convert   -c ${STRAIN}/REF-to-${STRAIN}.chain -i ${GTF} -f gtf -o ${STRAIN}/${STRAIN}.gtf

g2gtools gtf2db                -i ${STRAIN}/${STRAIN}.gtf -o ${STRAIN}/${STRAIN}.gtf.db
g2gtools extract --transcripts -i ${STRAIN}/${STRAIN}.fa -db ${STRAIN}/${STRAIN}.gtf.db &gt; ${STRAIN}/${STRAIN}.transcripts.fa
g2gtools extract --genes       -i ${STRAIN}/${STRAIN}.fa -db ${STRAIN}/${STRAIN}.gtf.db &gt; ${STRAIN}/${STRAIN}.genes.fa
g2gtools extract --exons       -i ${STRAIN}/${STRAIN}.fa -db ${STRAIN}/${STRAIN}.gtf.db &gt; ${STRAIN}/${STRAIN}.exons.fa</code></pre>
<pre><code> g2gtools --help

 Usage: g2gtools [OPTIONS] COMMAND [ARGS]...

 g2gtools

╭─ Options ─────────────────────────────────────────────────────────────────────────────────────────────────╮
│ --version                                                                                                 │
│ --install-completion        [bash|zsh|fish|powershell|pwsh]  Install completion for the specified shell.  │
│                                                              [default: None]                              │
│ --show-completion           [bash|zsh|fish|powershell|pwsh]  Show completion for the specified shell, to  │
│                                                              copy it or customize the installation.       │
│                                                              [default: None]                              │
│ --help                                                       Show this message and exit.                  │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────╯
╭─ Commands ────────────────────────────────────────────────────────────────────────────────────────────────╮
│ convert           Convert coordinates of BAM|SAM|GTF|GFF|BED files                                        │
│ extract           Extract subsequence from a fasta file given a region                                    │
│ fasta-format      Reformat a Fasta file                                                                   │
│ gtf2db            Convert a GTF file to a G2G DB file                                                     │
│ parse-region      Parse a region from the command line.  Useful for verifying regions.                    │
│ patch             Patch SNPs onto the reference sequence                                                  │
│ transform         Incorporate indels onto the input sequence                                              │
│ vcf2vci           Create VCI file from VCF file(s)                                                        │
│ vci-query         Query a VCI file.                                                                       │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────╯</code></pre>
<p>For each strain the following is produced:</p>
<pre><code><span><span class="va">A_J.39.exons.fa</span></span>
<span><span class="va">A_J.39.fa</span></span>
<span><span class="va">A_J.39.fa.fai</span></span>
<span><span class="va">A_J.39.genes.fa</span></span>
<span><span class="va">A_J.39.gtf</span></span>
<span><span class="va">A_J.39.gtf.db</span></span>
<span><span class="va">A_J.39.gtf.unmapped</span></span>
<span><span class="va">A_J.39.patched.fa</span></span>
<span><span class="va">A_J.39.patched.fa.fai</span></span>
<span><span class="va">A_J.39.transcripts.fa</span></span>
<span><span class="va">A_J.39.vci.gz</span></span>
<span><span class="va">A_J.39.vci.gz.tbi</span></span></code></pre>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul><li>Use the default <code>nextflow</code> pipeline if working with DO
mice on sumner2.</li>
<li>You can query the VCFs and get the strain names using
<code>tabix</code>.</li>
<li>You can specify another set of strains and use the default reference
genome.</li>
<li>Advanced users can run the pipeline line-by-line and specify
different options.</li>
</ul></div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/installation.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/prepare-emase.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/installation.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Installing GBRS
        </a>
        <a class="chapter-link float-end" href="../instructor/prepare-emase.html" rel="next">
          Next: Prepare EMASE... 
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/dmgatti/GBRS_Tutorial/edit/main/episodes/prepare-pseudo-reference.md" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/dmgatti/GBRS_Tutorial/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/dmgatti/GBRS_Tutorial/" class="external-link">Source</a></p>
				<p><a href="https://github.com/dmgatti/GBRS_Tutorial/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:dan.gatti@jax.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.4" class="external-link">sandpaper (0.16.4)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.5" class="external-link">pegboard (0.7.5)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.2" class="external-link">varnish (1.0.2)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://dmgatti.github.io/GBRS_Tutorial/instructor/prepare-pseudo-reference.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries, RNA-Seq, genetic diversity",
  "name": "Preparing GBRS Pseudo-References",
  "creativeWorkStatus": "active",
  "url": "https://dmgatti.github.io/GBRS_Tutorial/instructor/prepare-pseudo-reference.html",
  "identifier": "https://dmgatti.github.io/GBRS_Tutorial/instructor/prepare-pseudo-reference.html",
  "dateCreated": "2023-08-10",
  "dateModified": "2024-03-27",
  "datePublished": "2024-04-23"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

